lib/saluki-components/src/transforms/dogstatsd_mapper/mod.rs

```diff
@@ -0,0 +1,353 @@
+    async fn build(&self) -> Result<Box<dyn SynchronousTransform + Send>, GenericError> {
+        let metric_mapper = self
+            .dogstatsd_mapper_profiles
+            .build(self.context_string_interner_bytes)?;
+        Ok(Box::new(DogstatsDMapper { metric_mapper }))
+    }
+}
+
+impl MemoryBounds for DogstatsDMapperConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        builder
+            .minimum()
+            // Capture the size of the heap allocation when the component is built.
+            .with_single_value::<DogstatsDMapper>("component struct")
+            // We also allocate the backing storage for the string interner up front, which is used by our context
+            // resolver.
+            .with_fixed_amount("string interner", self.context_string_interner_bytes.as_u64() as usize);
+    }
+}
+
+pub struct DogstatsDMapper {
+    metric_mapper: MetricMapper,
+}
+
+impl SynchronousTransform for DogstatsDMapper {
+    fn transform_buffer(&mut self, event_buffer: &mut FixedSizeEventBuffer) {
+        for event in event_buffer {
+            if let Some(metric) = event.try_as_metric_mut() {
+                if let Some(new_context) = self.metric_mapper.map(
+                    metric.context().name(),
+                    metric.context().tags().to_owned(),
+                    metric.context().origin_tags().clone(),
+                ) {
+                    *metric.context_mut() = new_context;
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use bytesize::ByteSize;
+    use saluki_context::Context;
+    use saluki_event::metric::Metric;
+    use stringtheory::MetaString;
+
+    use super::{MapperProfileConfigs, MetricMapper};
+
+    fn mapper(json_data: &str) -> MetricMapper {
+        let mpc: MapperProfileConfigs = serde_json::from_str(json_data).unwrap();
+        let context_string_interner_bytes = ByteSize::kib(64);
+        mpc.build(context_string_interner_bytes).unwrap()
+    }
+
+    #[test]
+    fn test_mapper_wildcard() {
+        let json_data = r#"[{"name":"my_custom_metric_profile","prefix":"custom_metric.","mappings":[{"match":"custom_metric.process.*.*","match_type":"wildcard","name":"custom_metric.process","tags":{"tag_key_1":"$1","tag_key_2":"$2"}}]}]"#;
+        let mut mapper = mapper(json_data);
+        let context = Context::from_static_parts("custom_metric.process.value_1.value_2", &["foo:bar", "baz"]);
+        let input_metric_name = context.name().clone();
+        let metric = Metric::counter(context, 1.0);
+        let existing_tags = metric.context().tags();
+        let context = mapper
+            .map(
+                &input_metric_name,
+                existing_tags.clone(),
+                metric.context().origin_tags().clone(),
+            )
+            .unwrap();
+        let expected_metric_name = MetaString::from_static("custom_metric.process");
```

Comment:
FWIW, I don't _think_ you should need to make this `MetaString` explicitly: `MetaString` dereferences to `&str`, which should mean it's fine to just do `assert_eq!(context.name(), "custom_metric.process");`

Did you try that and it didn't happen to work? Genuinely curious because that's how I would _want_ it to work, and if it doesn't, I want to tweak `MetaString` so that it does work.